src/agentscope/agents/web_voyager_agent.py

```diff
@@ -0,0 +1,411 @@
+                # TODO change fail obs
+                curr_msg = Msg(
+                    name="user",
+                    content=fail_obs,
+                    role="user",
+                )
+                self.memory.add(curr_msg)
+
+            messages = self._clip_message_and_obs(self.memory.get_memory())
+
+            formated_messages = self.model.format(messages)
+            gpt_4v_res = self.model(formated_messages).text
+            self.speak(gpt_4v_res)
+
+            self.memory.add(
+                Msg(name="assistant", content=gpt_4v_res, role="assistant"),
+            )
+
+            # extract action info
+            try:
+                assert "Thought:" in gpt_4v_res and "Action:" in gpt_4v_res
+            except AssertionError as e:
+                logger.error(e)
+                fail_obs = "Format ERROR: Both 'Thought' and 'Action' should be included in your reply."  # noqa
+                continue
+
+            # bot_thought = re.split(pattern, gpt_4v_res)[1].strip()
+            bot_action = re.split(pattern, gpt_4v_res)[2].strip()
+            action_key, info = self._extract_action(bot_action)
+
+            fail_obs = ""
+            warn_obs = ""
+            try:
+                # TODO how to prevent bad auto page navigation?
+
+                if action_key == "click":
+                    click_ele_number = int(info[0])
+                    ele_tag_name = web_ele_fields[click_ele_number]["tag_name"]
+                    ele_type = web_ele_fields[click_ele_number]["type"]
+
+                    self.browser.click(click_ele_number)
+
+                    # TODO what to do when encounter PDF file
+
+                    if ele_tag_name == "button" and ele_type == "submit":
+                        time.sleep(10)
+                    time.sleep(5)
+
+                elif action_key == "wait":
+                    time.sleep(5)
+
+                elif action_key.startswith("type"):
+                    warn_obs = ""
+                    type_content = info["content"]
+                    type_ele_number = int(info["number"])
+
+                    ele_tag_name = web_ele_fields[type_ele_number]["tag_name"]
+                    ele_type = web_ele_fields[type_ele_number]["type"]
+                    if (ele_tag_name not in ("input", "textarea")) or (
+                        ele_tag_name == "input"
+                        and ele_type
+                        not in ["text", "search", "password", "email", "tel"]
+                    ):
+                        warn_obs = f"note: The web element you're trying to type may not be a textbox, and its tag name is <{ele_tag_name}>, type is {ele_type}."  # noqa
+
+                    self.browser.type(
+                        type_ele_number,
+                        type_content,
+                        # submit=(action_key == "typesubmit"),
+                        submit=True,
+                    )
+
+                    if "wolfram" in self.task_web:
```

Comment:
What is this line for?

Reply:
Deleted. I will find a another way to make sure the page is fully loaded before the agent proceed, as certain web page will require more time to load, .